### PR TITLE
DiscreteVariable: Change `value` from list to tuple

### PR DESCRIPTION
**Pull request by janezd**
##### Issue

It is understood<sup>[by whom?]</sup> that `DiscreteVariable.values` shouldn't be modified.

##### Description of changes

It is enforced that `DiscreteVariable.values` can't be modified: it is made a read-only property, and changed from list to tuple.

##### Includes

- [X] Code changes
- [X] Tests

### CODE REVIEW

**Comment by janezd** [comment]
The last commit, https://github.com/biolab/orange3/pull/4461/commits/ed4fbc18208f6d707a6c9fa66fc3de6563c45bb7, is a temporary backward compatibility hack that we can also omit if we think it's too ugly. @markotoplak, @lanzagar, @VesnaT?

File: Orange/data/variable.py [hunk 1]
```diff
@@ -560,6 +560,24 @@ def copy(self, compute_value=None, *, name=None, **kwargs):
...
+                "DiscreteVariable.values is a tuple; "
+                "support for adding a list will be dropped in Orange 3.27",
+                DeprecationWarning)
+            other = tuple(other)
+        return super().__add__(other)
+
+    def copy(self):
+        warnings.warn(
+            "DiscreteVariable.values is a tuple;"
+            "method copy  will be dropped in Orange 3.27",
```
Comment (lanzagar) [inline Orange/data/variable.py]: This is the one I am most proud of catching:
2 spaces after copy

File: Orange/data/variable.py [hunk 2]
```diff
@@ -744,25 +766,25 @@ def repr_val(self, val):
...
-        __dict__.pop("values")
+        __dict__.pop("_values")
         return make_variable, (self.__class__, self._compute_value, self.name,
-                               self.values, self.ordered), \
+                               self._values, self.ordered), \
             __dict__
 
     def copy(self, compute_value=None, *, name=None, values=None, **_):
-        if values is not None and len(values) != len(self.values):
+        if values is not None and len(values) != len(self._values):
```
Comment (lanzagar) [inline Orange/data/variable.py]: Why was this change necessary? `self.values` being a property should still work right?

There are similar cases in other places as well...

File: doc/data-mining-library/source/tutorial/data.rst [hunk 3]
```diff
@@ -39,13 +39,13 @@ The rest of the table gives the data. Note that there are 5 instances in our tab
 
 Note that for the file name no suffix is needed, as Orange checks if any files in the current directory are of a readable type. The call to ``Orange.data.Table`` creates an object called ``data`` that holds your dataset and information about the lenses domain:
 
-    >>> data.domain.attributes
-    (DiscreteVariable('age', values=['pre-presbyopic', 'presbyopic', 'young']),
-     DiscreteVariable('prescription', values=['hypermetrope', 'myope']),
-     DiscreteVariable('astigmatic', values=['no', 'yes']),
-     DiscreteVariable('tear_rate', values=['normal', 'reduced']))
+    >>> data.domain.attribute()
```
Comment (lanzagar) [inline doc/data-mining-library/source/tutorial/data.rst]: ??

File: Orange/data/tests/test_variable.py [hunk 4]
```diff
@@ -194,22 +196,35 @@ def test_repr(self):
         var = DiscreteVariable.make("a", values=["F", "M"])
         self.assertEqual(
             repr(var),
-            "DiscreteVariable(name='a', values=['F', 'M'])")
+            "DiscreteVariable(name='a', values=('F', 'M'))")
```
[the remark this answers is not in the document]
Comment (janezd) [reply Orange/data/tests/test_variable.py]: Mhm, 286 changes. :)

https://github.com/biolab/orange3/pull/4461/commits/3e91b86e41cdff284f3fe337096d71a228c4aad3

I left those where a list comprehension is assigned to a value. For practicality.

File: Orange/data/tests/test_variable.py [hunk 5]
```diff
@@ -194,22 +196,35 @@ def test_repr(self):
...
-            "DiscreteVariable(name='a', values=['1', '2', '3', '4', '5', '6', '7'])")
+            "DiscreteVariable(name='a', values=('1', '2', '3', '4', '5', '6', '7'))")
 
     def test_no_nonstringvalues(self):
         self.assertRaises(TypeError, DiscreteVariable, "foo", values=["a", 42])
         a = DiscreteVariable("foo", values=["a", "b", "c"])
         self.assertRaises(TypeError, a.add_value, 42)
 
+    def test_tuple_list_warning(self):
+        if LooseVersion(Orange.__version__) >= LooseVersion("3.27"):
```
[the remark this answers is not in the document]
Comment (janezd) [reply Orange/data/tests/test_variable.py]: I copied this from somewhere. Fixed, thanks.

File: Orange/data/variable.py [hunk 6]
```diff
@@ -560,6 +560,24 @@ def copy(self, compute_value=None, *, name=None, **kwargs):
...
+                DeprecationWarning)
+            other = tuple(other)
+        return super().__add__(other)
+
+    def copy(self):
+        warnings.warn(
+            "DiscreteVariable.values is a tuple;"
+            "method copy  will be dropped in Orange 3.27",
+            DeprecationWarning)
+        return tuple(self)
```
[the remark this answers is not in the document]
Comment (janezd) [reply Orange/data/variable.py]: You're right. Done.

File: Orange/data/variable.py [hunk 7]
```diff
@@ -560,6 +560,24 @@ def copy(self, compute_value=None, *, name=None, **kwargs):
...
 
+class TupleList(tuple):
+    def __add__(self, other):
+        if isinstance(other, list):
+            warnings.warn(
+                "DiscreteVariable.values is a tuple; "
+                "support for adding a list will be dropped in Orange 3.27",
+                DeprecationWarning)
+            other = tuple(other)
+        return super().__add__(other)
```
[the remark this answers is not in the document]
Comment (janezd) [reply Orange/data/variable.py]: Same.

File: Orange/data/variable.py (hunk 1 above)
Comment (janezd) [reply Orange/data/variable.py]: Congrats!

File: Orange/data/variable.py (hunk 2 above)
Comment (janezd) [reply Orange/data/variable.py]: I can't decide. I changed it back.

File: doc/data-mining-library/source/tutorial/data.rst (hunk 3 above)
Comment (janezd) [reply doc/data-mining-library/source/tutorial/data.rst]: PyCharm froze while I was typing, and I couldn't find what changed, so I assumed I didn't change anything.
